src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs

```diff
@@ -424,8 +436,65 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
+                        {
+                            if (includeExpression.TargetResourceType != null)
+                            {
+                                resourceId = Model.GetResourceTypeId(includeExpression.TargetResourceType);
+                            }
+                            else if (includeExpression.ReferenceSearchParameter?.TargetResourceTypes?.Count > 0)
+                            {
+                                resourceId = Model.GetResourceTypeId(includeExpression.ReferenceSearchParameter.TargetResourceTypes.ToList().First());
+                            }
+                        }
+
                         delimited.BeginDelimitedElement().Append(VLatest.ReferenceSearchParam.ResourceTypeId, table)
-                            .Append(" = ").Append(Parameters.AddParameter(VLatest.ReferenceSearchParam.ResourceTypeId, Model.GetResourceTypeId(includeExpression.ResourceType)));
+                        .Append(" = ").Append(Parameters.AddParameter(VLatest.ReferenceSearchParam.ResourceTypeId, resourceId));
+
+                        // Get FROM ctes
+                        string fromCte = _cteMainSelect;
+                        if (includeExpression.Iterate)
+                        {
+                            // Include Iterate
+                            if (!includeExpression.Reversed)
+                            {
+                                // _include:iterate may appear without a preceding _include, in case of circular reference
+                                // On that case, the fromCte is _cteMainSelect
+                                if (TryGetIncludeCtes(includeExpression.SourceResourceType, out _includeFromCteIds))
+                                {
+                                    fromCte = _includeFromCteIds[++_curFromCteIndex];
+                                }
+                            }
+
+                            // RevInclude Iterate
+                            else
+                            {
+                                if (includeExpression.TargetResourceType != null)
+                                {
+                                    if (TryGetIncludeCtes(includeExpression.TargetResourceType, out _includeFromCteIds))
+                                    {
+                                        fromCte = _includeFromCteIds[++_curFromCteIndex];
+                                    }
+                                }
+                                else if (includeExpression.ReferenceSearchParameter?.TargetResourceTypes != null)
+                                {
+                                    // Assumes TargetResourceTypes is of length 1. Otherwise, a BadRequest would have been thrown earlier for _revinclude:iterate
+                                    List<string> fromCtes;
+                                    var targetType = includeExpression.ReferenceSearchParameter.TargetResourceTypes.First();
```

Comment:
Should we use `.Single()` to enforce this assumption?

Reply:
We could, but as I answered to Elad, if there are multiple target types and one is not specified, a BadRequest is thrown earlier and this is validated in an E2E test. We have E2E test to validate that.

Moreover, if you enforce it here (SQLQueryGenerator class), we will need to throw a BadRequest for throw without catching which will end up in an internal error. Both are not ideal IMO. We assume a valid request by the time we get to SqlQueryGenerator. Notice that we don;t  throw BadRequest from any code path in this class.

WDYT?

Reply:
Agree a Bad Request should be thrown before we get here. I'm happy to keep First(), but perhaps lets Log or Assert something if the expectation is not met? If things change upstream in future somehow, this has the potential to return the wrong information

Reply:
If things change in the future, a test will fail (I added it myself).
But I agree with you, logging is a good way to address it. 
The problem though is that there's no logger in SqlQueryGenerator class, so either one of the two:
1. Add logger to SqlQueryGenerator class (dependency injection) 
2. Through and catch at the SqlServerSearchService  and log 

I don;t think any of the option is good.
WDYT?

Reply:
If we have a test to validate, happy to leave it with this for now